docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,546 @@
+
+2. Channel Content (streaming)
+
+3. Bytes (non-streaming)
+
+4. Empty (non-streaming)
+
+And all of those are handled differently. Here's how each one should be handled (in the same order as above):
+
+1. Compression will be applied for streaming HTTP bodies
+
+2. Compression will be applied for streaming HTTP bodies
+
+3. Compression will be applied for non streaming HTTP bodies
+
+4. No compression is applied
+
+Here is how it works in more detail:
+
+```kotlin
+val request = context.protocolRequest.toBuilder()
+
+request.body = when(request.body) {
+    is HttpBody.SourceContent -> // Apply compression for streaming HTTP bodies
+    is HttpBody.ChannelContent -> // Apply compression for streaming HTTP bodies
+    is HttpBody.Bytes -> // Apply compression for non streaming HTTP bodies
+    is HttpBody.Empty -> // Return as is (no compression is applied)
+    else -> throw ClientException("HttpBody type is not supported")
+}
+
+request.build()
+```
+
+## Compression Algorithms
+Given that Kotlin is multi-platform there should be multiple "implementations" of each compression algorithm. As of the time of writing this doc (11/07/2023) Gzip is the only supported compression algorithm by the server side and the only one we will be supporting for
+the SDK. For the time being, only Gzip for JVM will be supported specifically. This will be done by using the following packages from the Java standard library: `GZIPInputStream` and `GZIPOutputStream`. The other KMP targets are TODO's, and we have plans of using the CRT implementation for those.
+
+We'll be able to add new ones as they start becoming available/required using the compression algorithm interface we've written and
+by adding them to the list of supported algorithms. The same goes for non JVM KMP targets, these can be added by filling in the `actual` declarations for the target.
+
+#### Interfaces
+This interface will be used to represent a compression algorithm.
+
+```kotlin
+/**
+ * Represents a compression algorithm to be used for compressing request payloads on qualifying operations
+ */
+public interface CompressionAlgorithm {
+    /**
+     * The ID of the checksum algorithm
+     */
+    public fun algorithmId(): String
+
+    /**
+     * Compresses a non-stream payload
+     */
+    public fun compress(content: ByteArray?): ByteArray
+
+    /**
+     * Compresses a stream payload
+     */
+    public suspend fun getCompressionStream(stream: ByteStream?): ByteStream
+}
+```
+
+The `algorithmId()` will be used to match a supported `CompressionAlgorithm` to a requested compression algorithm from the trait. The `compress()` function will be used to compress HTTP bodies that are not streaming. And the `getCompressionStream()` function will do the same but for streaming HTTP bodies.
+
+#### Compression
+Since only Gzip for JVM will be supported at the time that is all that is in the design doc. In the future more `actual` implementations may be added here if deemed necessary.
```

Comment:
you can remove this sentence, duplicated description of algorithms not implemented